src/libmeasurement_kit/nettests/base_test.cpp

```diff
@@ -1,175 +0,0 @@
-}
-
-BaseTest &BaseTest::set_output_filepath(std::string s) {
-    runnable->output_filepath = s;
-    return *this;
-}
-
-BaseTest &BaseTest::set_error_filepath(std::string s) {
-    runnable->logger->set_logfile(s);
-    return *this;
-}
-
-BaseTest &BaseTest::add_annotation(std::string key, std::string value) {
-    runnable->annotations[key] = value;
-    return *this;
-}
-
-BaseTest &BaseTest::set_option(std::string key, std::string value) {
-    runnable->options[key] = value;
-    return *this;
-}
-
-BaseTest &BaseTest::on_entry(Callback<std::string> &&cb) {
-    runnable->entry_cb = cb;
-    return *this;
-}
-
-BaseTest &BaseTest::on_begin(Callback<> &&cb) {
-    runnable->begin_cb = cb;
-    return *this;
-}
-
-BaseTest &BaseTest::on_end(Callback<> &&cb) {
-    runnable->end_cbs.push_back(cb);
-    return *this;
-}
-
-BaseTest &BaseTest::on_destroy(Callback<> &&cb) {
-    runnable->destroy_cbs.push_back(cb);
-    return *this;
-}
-
-BaseTest &BaseTest::on_overall_data_usage(Callback<DataUsage> &&cb) {
-    runnable->data_usage_cb = std::move(cb);
-    return *this;
-}
-
-static void start_internal_(SharedPtr<Runnable> &&r, std::promise<void> *promise,
-                            Callback<> &&callback) {
-    // Note:
-    //
-    // 1. we wait in the default tasks queue for our turn: this guarantees
-    //    that "big" operations like tests and tasks are sequenced.
-    //
-    // 2. we move all context inside the first lambda callback so that the
-    //    thread that will run I/O has exclusive ownership.
-    //
-    // 3. run_with_initial_event() is blocking: this means that the lifecycle
-    //    of `r` is guaranteed to be as long as the one of the reactor.
-    //
-    // 4. we _explicitly_ leave the reactor loop on success, but, even in
-    //    case the final event is not reached because of a bug, the reactor
-    //    is anyway going to exit the loop when it is out of events.
-    //
-    // 5. the `promise`, if present, allows to make the test synchronous,
-    //    while the callback allows to make it asynchronous.
-    assert(!r->reactor);
-    // Note: making a copy of the logger and moving `r` inside the closure
-    // to avoid creating a reference loop. I like to think at references as
-    // ropes and as reference loops as yards of rope around my neck. This
-    // is to say, dammit!, I/we should make sure we have more unique owner-
-    // ship in measurement-kit, to reduce cases like this one.
-    auto logger = r->logger;
-    Worker::default_tasks_queue()->call_in_thread(logger,
-          [ r = std::move(r), promise, callback = std::move(callback) ] {
-              r->reactor = Reactor::make();
-              r->reactor->run_with_initial_event([&]() {
-                  r->begin([&](Error) {
-                      r->end([&](Error) {
-                          r->reactor->stop();
-                          if (callback) {
-                              callback();
-                          }
```

Comment:
FWIW, this code has the bug that, if we leave the reactor for some reason, the final callback isn't called.